# CRISPR

### ADAM2

R package for identification of essential genes using genome-wide CRISPR-Cas9 screens, <https://github.com/DepMap-Analytics/ADAM2>.

### CHOPCHOP

CHOPCHOP is a python script that allows quick and customizable design of guide RNA.
We support selecting target sites for CRISPR/Cas9, CRISPR/Cpf1, TALEN and NICKASE with wide
range of customization. We even support Cas13 for isoform targeting.

<https://chopchop.cbu.uib.no/>

<https://bitbucket.org/valenlab/chopchop/src/master/>

### CRISPOR

<http://crispor.tefor.net/>, design, evaluate and clone guide sequences for the CRISPR/Cas9 system.

### CRISPR design

<http://crispr.dbcls.jp/>

### CRISPResso2

<https://github.com/pinellolab/CRISPResso2>

### DeepSpCas9

<http://deepcrispr.info/DeepSpCas9/>

Kim HK, et al. (2019). SpCas9 activity prediction by DeepSpCas9, a deep learning–based model with high generalization performance.  *Sci. Adv.* 2019;5: eaax9249.

### FORECasT

Scripts for processing and predicting CRISPR/Cas9-generated mutations.

<https://partslab.sanger.ac.uk/FORECasT>

Allen F, et al. (2019). Predicting the mutations generated by repair of Cas9-induced double-strand breaks. *Nat Biotechnol* 37: 64–72

## GPP Web portal

<https://portals.broadinstitute.org/gpp/public/analysis-tools/sgrna-design>

### inDelphi

inDelphi is a machine learning algorithm that is aimed to assist scientists using CRISPR.

<http://indelphi.giffordlab.mit.edu/>

### OpenCRISPR

<https://github.com/Profluent-AI/OpenCRISPR>

Ruffolo, J. A., Nayfach, S., Gallagher, J., Bhatnagar, A., Beazer, J., Hussain, R., Russ, J., Yip, J., Hill, E., Pačesa, M., Meeske, A. J., Cameron, P., & Madani, A. (2024). Design of highly functional genome editors by modeling the universe of CRISPR-Cas sequences. bioRxiv. <https://doi.org/10.1101/2024.04.22.590591>

### SelfTarget

<https://github.com/felicityallen/SelfTarget>

### SPROUT

<https://zou-group.github.io/SPROUT>

<https://github.com/amirmohan/SPROUT>

Leenay RT, et al. (2019). Large dataset enables prediction of repair after CRISPR–Cas9 editing in primary T cells. *Nat Biotechnol* 
