# Long-read sequencing

## Company sites

- AWS, <https://github.com/aws>
- ONT, <https://nanoporetech.com/>
- PacBio, <https://www.pacb.com/>

## Software

- AWS CLI Command Reference, <https://docs.aws.amazon.com/cli/latest/>
- cuteFC, <https://github.com/Meltpinkg/cuteFC>
- cuteSV, <https://github.com/tjiangHIT/cuteSV>
- GAS power calculator, <https://csg.sph.umich.edu/abecasis/gas_power_calculator/index.html>
- minimap2, <https://github.com/lh3/minimap2>
- ONT, <https://github.com/nanoporetech>
- pbmm2, <https://github.com/PacificBiosciences/pbmm2>
- pbsv, <https://github.com/PacificBiosciences/pbsv>
- RTG tools, <https://github.com/RealTimeGenomics/rtg-tools>
- Sniffles, <https://github.com/fritzsedlazeck/Sniffles>
- Stragli, <https://github.com/philres/straglr>
- Truvari, <https://github.com/ACEnglish/truvari>

## Data sources

- EPI2ME, <https://epi2me.nanoporetech.com/dataindex/>
- GIAB, <https://ftp.ncbi.nlm.nih.gov/giab/ftp/>. The ONT PromethION dataset of the HG002 sample is here,
<https://ftp.ncbi.nlm.nih.gov/giab/ftp/data/AshkenazimTrio/HG002_NA24385_son/UCSC_Ultralong_OxfordNanopore_Promethion/>
- IGSR, <https://www.internationalgenome.org/>

## References

***Genome Res*** 2024, <https://genome.cshlp.org/content/34/11.toc>

Bolognini D, et al. VISOR: a versatile haplotype-aware structural variant simulator for short- and long-read sequencing, ***Bioinformatics***, 36(4), 1267–1269 (2020), <https://doi.org/10.1093/bioinformatics/btz719>

Jiang T, et al. Long-read-based human genomic structural variation detection with cuteSV. ***Genome Biol*** 21, 189 (2020). <https://doi.org/10.1186/s13059-020-02107-y>. The paper has scripts for various software.

Smolka M, et al. Detection of mosaic and population-level structural variants with Sniffles2. ***Nat Biotechnol*** 42, 1571–1580 (2024). <https://doi.org/10.1038/s41587-023-02024-y>

Xia LC, et al. SVEngine: an efficient and versatile simulator of genome structural variations with features of cancer clonal evolution, ***GigaScience***, 7(7), giy081 (2018), <https://doi.org/10.1093/gigascience/giy081>

Yuan N, Jia P. Comprehensive assessment of long-read sequencing platforms and calling algorithms for detection of copy number variation, ***Briefings in Bioinformatics***, Volume 25, Issue 5, September 2024, bbae441, <https://doi.org/10.1093/bib/bbae441>. The paper includes an overview of software.

Zook JM, et al. A robust benchmark for detection of germline large deletions and insertions. ***Nat Biotechnol*** 38, 1347–1355 (2020). <https://doi.org/10.1038/s41587-020-0538-8>
