# Single-cell analysis

Bioconductor branch, <http://bioconductor.org/packages/release/BiocViews.html#___SingleCell>.

[benchmarks](https://github.com/tabdelaal/scRNAseq_Benchmark/tree/snakemake_and_docker)

Abdelaal et al. A comparison of automatic cell identification methods for single-cell RNA sequencing data. *Genome Biol* (2019) 20:194,
<https://doi.org/10.1186/s13059-019-1795-z>

[Single cell RNA-seq data analysis bundle](https://github.com/haniffalab/FCA_liver).

Popescu D-M et al. Decoding human fetal liver haematopoiesis. *Nature* <https://doi.org/10.1038/s41586-019-1652-y>.

## BayesPrism

GitHub: <https://github.com/Danko-Lab/BayesPrism>

Chu, T., Wang, Z., Pe’er, D. & Danko, C.G. Cell type and gene expression deconvolution with BayesPrism enables Bayesian integrative analysis across bulk and single-cell RNA sequencing in oncology. *Nature Cancer* 3, 505-517 (2022).

## CReSCENT

Fullname: CanceR Single Cell ExpressioN Toolkit

Web: <https://crescent.cloud/>

Mohanraj, S., et al. (2020). "CReSCENT: CanceR Single Cell ExpressioN Toolkit." *Nucleic Acids Research*.

## cytoSCAPE

GitHub: <https://github.com/digitalcytometry/cytospace>; Web: <https://cytospace.stanford.edu/>

## DENDRO

Web: <https://github.com/zhouzilu/DENDRO>

Zhou Z, et al. DENDRO: genetic heterogeneity profiling and subclone detection by single-cell RNA sequencing. *Genome Biol* 21, 10 (2020). <https://doi.org/10.1186/s13059-019-1922-x>

## SCANPY

Web: <https://github.com/theislab/Scanpy>

## SCENIC

Web: <https://aertslab.org/#scenic> and <https://pyscenic.readthedocs.io/en/latest/>.

## scGLUE

Web: <https://scglue.readthedocs.io/en/latest/index.html>

GitHub: <https://github.com/gao-lab/GLUE>

Cao ZJ, Gao G. Multi-omics single-cell data integration and regulatory inference with graph-linked embedding. *Nat Biotechnol* 40, 1458–1466 (2022). <https://doi.org/10.1038/s41587-022-01284-4>

## scGPT

Web: <https://github.com/bowang-lab/scGPT>

Cui, H., Wang, C., Maan, H. et al. scGPT: toward building a foundation model for single-cell multi-omics using generative AI. *Nat Methods* 21, 1470–1480 (2024). <https://doi.org/10.1038/s41592-024-02201-0>

## scMAGeCK

Web: <https://bitbucket.org/weililab/scmageck/src/master/>

Yang L, et al. scMAGeCK links genotypes with multiple phenotypes in single-cell CRISPR screens. *Genome Biol* 21, 19 (2020). <https://doi.org/10.1186/s13059-020-1928-4>

## Seurat

Web: <https://satijalab.org/seurat/>

Stuart T, et al. Comprehensive Integration of Single-Cell Data. *Cell*. 2019;177(7):1888–1902.e21. doi:10.1016/j.cell.2019.05.031

## Signac

Web: <https://cloud.r-project.org/web/packages/Signac>

Stuart, T., Srivastava, A., Madad, S., Lareau, C.A. & Satija, R. Single-cell chromatin state analysis with Signac. *Nat Methods* 18, 1333-1341 (2021).

## souporcell

Web: <https://github.com/wheaton5/souporcell>

Heaton H, et al. Souporcell: robust clustering of single-cell RNA-seq data by genotype without reference genotypes. *Nat Methods* (2020). <https://doi.org/10.1038/s41592-020-0820-1>

## squidpy

Web: <https://squidpy.readthedocs.io/en/latest/>.

Palla, G. et al. Squidpy: a scalable framework for spatial omics analysis. *Nat Methods* (2022).
